linalg/AmgXSolver.cpp

```diff
@@ -0,0 +1,846 @@
+// Copyright (c) 2010-2020, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+//Based on the work of:
+//Pi-Yueh Chuang, & Lorena A. Barba (2017).
+//AmgXWrapper: An interface between PETSc and the NVIDIA AmgX library. J.
+//Open Source Software, 2(16):280, doi:10.21105/joss.00280
+
+#include "../config/config.hpp"
+#include "AmgXSolver.hpp"
+#ifdef MFEM_USE_AMGX
+
+namespace mfem
+{
+
+// Initialize AmgXSolver::count to 0
+int AmgXSolver::count = 0;
+
+// Initialize AmgXSolver::rsrc to nullptr;
+AMGX_resources_handle AmgXSolver::rsrc = nullptr;
+
+// Implements AmgXSolver::AmgXSolver
+AmgXSolver::AmgXSolver(const std::string &modeStr, const std::string &cfgFile)
+{
+   Initialize_Serial(modeStr, cfgFile);
+}
+
+#ifdef MFEM_USE_MPI
+AmgXSolver::AmgXSolver(const MPI_Comm &comm,
+                       const std::string &modeStr, const std::string &cfgFile, int &nDevs)
+{
+   Initialize_MPITeams(comm, modeStr, cfgFile, nDevs);
+}
+
+AmgXSolver::AmgXSolver(const MPI_Comm &comm,
+                       const std::string &modeStr, const std::string &cfgFile)
+{
+   Initialize_ExclusiveGPU(comm, modeStr, cfgFile);
+}
+#endif
+
+AmgXSolver::~AmgXSolver()
+{
+   if (isInitialized) { finalize(); }
+}
+
+void AmgXSolver::Initialize_Serial(const std::string &modeStr,
+                                   const std::string &cfgFile)
+{
```

Comment:
Is it necessary to specify the configuration parameters by an input file? This could be inconvenient and restrictive.	It would give the user more flexibility to have a constructor allowing specification of the parameters in the code. Would that be possible with AmgX?